Fix inconsisteny between Python and JIT power operation

szewaiyuen7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62842

Differential Revision: [D30146115](https://our.internmc.facebook.com/intern/diff/D30146115)

szewaiyuen7 committed:
Fix inconsisteny between Python and JIT power operation

[ghstack-poisoned]

szewaiyuen7 committed:
Update on "Fix inconsisteny between Python and JIT power operation"




[ghstack-poisoned]

szewaiyuen7 commented:
@szewaiyuen7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30146115).

szewaiyuen7 committed:
Update on "Fix inconsisteny between Python and JIT power operation"


Differential Revision: [D30146115](https://our.internmc.facebook.com/intern/diff/D30146115)

[ghstack-poisoned]

szewaiyuen7 commented:
@szewaiyuen7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30146115).

zhxchen17 commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -912,7 +922,7 @@ RegisterOperators reg(
            int64_t a, b;
            pop(stack, a, b);
-           push(stack, pow(a, b));
+           push(stack, static_cast<double>(powWrapper(a, b)));
```
why do we need to add static_cast? I think implicit cast is sufficient here.

szewaiyuen7 replied:
I am not sure I understand why in this particular case we do not need explicit casting when all the above test cases require explicit casting. Would love to get a bit more details on your rationale.

The static_cast<double>(pow(a, b)) is passed in as one of the xx_op argument into the operator generation, and then passed directly into the push function. (e.g. https://github.com/pytorch/pytorch/blob/8c852de54d8b07c1fa546cbd400c093c4b2082d6/torch/csrc/jit/runtime/register_ops_utils.h#L712)

Therefore, I also add casting here to be consistent. 

If you can tell me more on what makes this case different, would be super helpful!

zhxchen17 replied:
IIUC from https://en.cppreference.com/w/cpp/numeric/math/pow, it seems if two inputs are all ints, the output is promoted to double automatically, so static_cast is not needed here.

szewaiyuen7 replied:
Makes a lot of sense, thank you!

zhxchen17 commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -893,13 +900,16 @@ RegisterOperators reg(
+         aten::pow,
+         static_cast<double>(powWrapper(a, b)),
+         float),
      DEFINE_FLOAT_COMPLEX_OP(aten::pow, pow(a, b), complex),
```
I tested in Python that if a we have 0 ** complex, and the real part of the complex number is negative, it also raises an error, could we add the check for this case?

szewaiyuen7 replied:
Good call, let me add it.

szewaiyuen7 replied:
I think Python prevents 0 to the power of any complex number, not just those where the real part is negative. For example, `print(0 ** complex(10, 1))` and `print(0 ** complex(-10, 1))` both give error `ZeroDivisionError: 0.0 to a negative or complex power.`

PyTorch on the other hand, return `0j` for `0 ** complex(10, 1)` and `(inf+nanj)` for `0 ** complex(-10, 1)`.

I think the discussion is more broadly whether PyTorch should allow 0 being raised to a complex number, which I do not have a definitive answer to. Therefore, it may make sense to create a separate issue to track the zero base complex exponent inconsistency and limit the scope of this diff to real numbers.

zhxchen17 commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -893,13 +900,16 @@ RegisterOperators reg(
+         float),
      DEFINE_FLOAT_COMPLEX_OP(aten::pow, pow(a, b), complex),
      DEFINE_SCALAR_BINARY_OP_AVOID_COLLISION(
          aten::pow,
```
Let's just wrap this case and the other ones altogether.

szewaiyuen7 replied:
I am not sure how to trigger this function call in Python and therefore, not sure how to test for this. Any thoughts?

zhxchen17 replied:
yeah I talked with @gmagogsfm and seems like this can be triggered when we have int ** complex, unfortunately this will crash right now because the complex case is not implemented. We can use the wrapper here and make a separate diff to implement the complex number case for scalar `binary` operations.

For the (int, int) -> int case, I don't think it's used anywhere, so we can either wrap it, or remove it.

szewaiyuen7 replied:
That makes sense, thank you!

zhxchen17 reviewed (commented):
Thanks for the work! some more comments.

szewaiyuen7 committed:
Update on "Fix inconsisteny between Python and JIT power operation"


Differential Revision: [D30146115](https://our.internmc.facebook.com/intern/diff/D30146115)

[ghstack-poisoned]

szewaiyuen7 commented:
@szewaiyuen7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30146115).

szewaiyuen7 commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -912,7 +922,7 @@ RegisterOperators reg(
            int64_t a, b;
            pop(stack, a, b);
-           push(stack, pow(a, b));
+           push(stack, powWrapper(a, b));
```
@zhxchen17, removing this operator led to internal tests failure: https://www.internalfb.com/diff/D30183946

I am going to keep this operator as a result.

szewaiyuen7 committed:
Update on "Fix inconsisteny between Python and JIT power operation"


Differential Revision: [D30146115](https://our.internmc.facebook.com/intern/diff/D30146115)

[ghstack-poisoned]

szewaiyuen7 commented:
@szewaiyuen7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30146115).

zhxchen17 reviewed (approved):
looks great!
